OrbitBase/include/OrbitBase/MainThreadExecutor.h

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORIBIT_BASE_MAIN_THREAD_EXECUTOR_H_
+#define ORIBIT_BASE_MAIN_THREAD_EXECUTOR_H_
+
+#include <memory>
+#include <thread>
+
+#include "OrbitBase/Action.h"
+
+// This class implements menchanizm for landing
+// actions to a main thread. As a general rule
+// waiting on sockets and processing should be
+// happening off main thread and main thread
+// should only be responsible for updating user
+// interface and models.
+//
+// Usage example:
+// /* A caller who wants to process something on the main thread,
+//  * note that this is not-blocking operation and will be processed
+//  * at some time in the future on the main thread.
+//  */
+//
+// manager->Schedule(CreateAction([data]{
+//   UpdateSomethingWith(data);
+// }));
+//
+// To consume events the main thread should pereodically call
+// manager->ConsumeActions();
+//
+class MainThreadExecutor {
+ public:
+  MainThreadExecutor() = default;
+  virtual ~MainThreadExecutor() = default;
+
+  // Schedules the action to be performed on main thread.
+  // TODO: should this be unique_ptr?
```

Comment:
Do you have a specific case in mind where ownership of actions needs to be shared? Otherwise I would have indeed suggested `unique_ptr`s.

Reply:
Not really.. going with unique_ptr until proven otherwise.